lib/puppet/type/file/content.rb

```diff
@@ -119,6 +120,11 @@ def insync?(is)
       result
     end
 
+    def property_matches?(current, desired)
+      return super unless [:mtime, :ctime].include?(checksum_type)
+      DateTime.parse(current) >= DateTime.parse(desired)
```

Comment:
I think `current` can be `nil` and/or `:absent`. The file `mode` property has some similar logic: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/file/mode.rb#L117-L122

I wonder what happens if you say `ensure => absent`, does the `property_matches?` method get called? For a "normal" type I would say no, I would expect only the providers' `destroy` method to be called. But the file type has intertwined properties and parameters, so I'm not sure.

Do we know that `current` and `desired` are always in a specific date-time format? I think these strings are generated via `File.stat(path).mtime`, with an implicit `Time#to_s` somewhere along the way. The `current` time is generated on the local system and `desired` on the (potentially) remote system. Hopefully ruby is consistent across platforms (windows agent & linux master) and versions (1.9.3 vs 2.2.0)?

Our `file_metadata` schema only says the checksum value is a string: https://github.com/puppetlabs/puppet/blob/master/api/schemas/file_metadata.json#L31-L40. Also it looks like the schema is missing some checksum types (`ctime` and `*lite`).